Applications/Spire/Source/Ui/FilteredDropDownMenu.cpp

```diff
@@ -0,0 +1,207 @@
+    on_item_activated(value);
+  });
+  m_list_selection_connection = m_menu_list->connect_value_selected_signal(
+    [=] (const auto& value) { on_item_selected(value); });
+  set_items(items);
+  m_menu_list->installEventFilter(this);
+}
+
+bool FilteredDropDownMenu::eventFilter(QObject* watched, QEvent* event) {
+  if(watched == m_menu_list) {
+    if(event->type() == QEvent::Show || event->type() == QEvent::Hide) {
+      m_menu_list->setFixedWidth(width());
+      update();
+    }
+  }
+  return TextInputWidget::eventFilter(watched, event);
+}
+
+void FilteredDropDownMenu::focusInEvent(QFocusEvent* event) {
+  TextInputWidget::focusInEvent(event);
+  if(event->reason() == Qt::FocusReason::MouseFocusReason) {
+    m_was_click_focused = true;
+  }
+}
+
+void FilteredDropDownMenu::keyPressEvent(QKeyEvent* event) {
+  if(event->key() == Qt::Key_Escape) {
+    setText(m_item_delegate.displayText(m_current_item));
+  } else if(event->key() == Qt::Key_Enter || event->key() == Qt::Key_Return) {
+    if(m_last_activated_item.isValid()) {
+      on_item_selected(m_last_activated_item);
+      return;
+    }
+    if(!text().isEmpty() && m_menu_list->isVisible()) {
+      auto item = std::move(m_menu_list->get_value(0));
+      if(item.isValid()) {
+        m_list_selection_connection.disconnect();
+        on_item_selected(item);
+      }
+    }
+    return;
+  }
+  TextInputWidget::keyPressEvent(event);
+}
+
+void FilteredDropDownMenu::mousePressEvent(QMouseEvent* event) {
+  TextInputWidget::mousePressEvent(event);
+  if(m_was_click_focused) {
+    selectAll();
+    m_was_click_focused = false;
+  }
+}
+
+void FilteredDropDownMenu::paintEvent(QPaintEvent* event) {
+  TextInputWidget::paintEvent(event);
+  if(!text().isEmpty()) {
+    auto item = std::move(m_menu_list->get_value(0));
+    if(item.isValid()) {
+      auto item_text = m_item_delegate.displayText(item);
+      if(item_text.startsWith(text(), Qt::CaseInsensitive) &&
+          item_text.length() != text().length()) {
+        item_text = item_text.remove(0, text().length());
+        draw_highlight(item_text);
+      }
+    }
+  }
+  if(m_menu_list->isActiveWindow()) {
+    draw_border(QColor("#4B23A0"));
+  }
+}
+
+connection FilteredDropDownMenu::connect_selected_signal(
+    const SelectedSignal::slot_type& slot) const {
+  return m_selected_signal.connect(slot);
+}
+
+void FilteredDropDownMenu::set_items(const std::vector<QVariant>& items) {
+  m_items = items;
+  m_menu_list->set_items(create_widget_items(m_items));
+  m_menu_list->setFixedWidth(width());
+}
+
+const std::vector<DropDownItem*> FilteredDropDownMenu::create_widget_items(
+    const std::vector<QVariant>& items) {
+  return create_widget_items(items, "");
+}
+
+const std::vector<DropDownItem*> FilteredDropDownMenu::create_widget_items(
+    const std::vector<QVariant>& items, const QString& filter_text) {
+  auto widget_items = std::vector<DropDownItem*>();
+  widget_items.reserve(items.size());
```

Comment:
Don't reserve capacity unless the filter is supposed to be true almost all of the times.